src/megaapi_impl.cpp

```diff
@@ -882,6 +930,194 @@ MegaHandle MegaNodePrivate::getOwner() const
 
 const unsigned int MegaApiImpl::MAX_SESSION_LENGTH = 64;
 
+
+MegaBackgroundMediaUploadPrivate::MegaBackgroundMediaUploadPrivate(MegaApiImpl* capi)
+    : api(capi)
+{
+    // generate fresh random encryption key/CTR IV for this file
+    api->client->rng.genblock(filekey, sizeof filekey);
+}
+
+MegaBackgroundMediaUploadPrivate::MegaBackgroundMediaUploadPrivate(const string& serialised, MegaApiImpl* capi)
+    : api(capi)
+{
+    CacheableReader r(serialised);
+    string mediapropertiesstr;
+    unsigned char expansions[8];
+    string fileattrstring; // fileattrstring is not serialized
+    if (!r.unserializebinary(filekey, sizeof(filekey)) ||
+        !r.unserializechunkmacs(chunkmacs) ||
+        !r.unserializestring(mediapropertiesstr) ||
+        !r.unserializestring(url) ||
+        !r.unserializeexpansionflags(expansions, 0))
+    {
+        LOG_err << "MegaBackgroundMediaUploadPrivate unserialization failed at field " << r.fieldnum;
+    }
+    else
+    {
+        mediaproperties = MediaProperties(mediapropertiesstr);
+    }
+}
+
+std::string MegaBackgroundMediaUploadPrivate::serialize()
+{
+    std::string s;
+    CacheableWriter w(s);
+    w.serializebinary(filekey, sizeof(filekey));
+    w.serializechunkmacs(chunkmacs);
+    w.serializestring(mediaproperties.serialize());
+    w.serializestring(url);
+    w.serializeexpansionflags(false, false);
+    return s;
+}
+
+
+MegaBackgroundMediaUploadPrivate::~MegaBackgroundMediaUploadPrivate()
+{
+}
+
+class MEGA_API EncryptFilePieceByChunks : public EncryptByChunks
```

Comment:
It's usually a good practice to keep declarations in header files and definitions in source files. Why is this one all together here? I'll fix it